src/core/lib/event_engine/windows/windows_listener.h

```diff
@@ -0,0 +1,126 @@
+      IOCP* iocp, AcceptCallback accept_cb,
+      absl::AnyInvocable<void(absl::Status)> on_shutdown,
+      std::unique_ptr<MemoryAllocatorFactory> memory_allocator_factory,
+      Executor* executor, const EndpointConfig& config);
+  ~WindowsEventEngineListener() override;
+  absl::StatusOr<int> Bind(const EventEngine::ResolvedAddress& addr) override;
+  absl::Status Start() override;
+
+ private:
+  friend class SinglePortSocketListener;
+  /// Responsible for listening on a single port.
+  class SinglePortSocketListener {
+   public:
+    ~SinglePortSocketListener();
+    // This factory will create a bound, listening WinSocket, registered with
+    // the listener's IOCP poller.
+    static absl::StatusOr<std::unique_ptr<SinglePortSocketListener>> Create(
+        WindowsEventEngineListener* listener, SOCKET sock,
+        EventEngine::ResolvedAddress addr);
+
+    // Two-stage initialization, allows creation of all bound sockets before the
+    // listener is started.
+    absl::Status Start();
+    absl::Status StartLocked() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+
+    // Accessor methods
+    EventEngine::ResolvedAddress listener_sockname() {
+      return listener_sockname_;
+    };
+    int port() { return port_; }
+    WinSocket* listener_socket() { return listener_socket_.get(); }
+
+   private:
+    SinglePortSocketListener(WindowsEventEngineListener* listener,
+                             LPFN_ACCEPTEX AcceptEx,
+                             grpc_core::OrphanablePtr<WinSocket> win_socket,
+                             int port, EventEngine::ResolvedAddress hostbyname);
+
+    // Bind a recently-created socket for listening
+    struct PrepareListenerSocketResult {
+      int port;
+      EventEngine::ResolvedAddress hostbyname;
+    };
+    static absl::StatusOr<PrepareListenerSocketResult> PrepareListenerSocket(
+        SOCKET sock, const EventEngine::ResolvedAddress& addr);
+
+    void OnAcceptCallbackImpl();
+
+    // The cached AcceptEx for that port.
+    LPFN_ACCEPTEX AcceptEx;
+    // This seemingly magic number comes from AcceptEx's documentation. each
+    // address buffer needs to have at least 16 more bytes at their end.
+    uint8_t addresses_[(sizeof(sockaddr_in6) + 16) * 2] = {};
+    // The parent listener
+    WindowsEventEngineListener* listener_;
+    // closure for socket notification of accept being ready
+    AnyInvocableClosure on_accept_;
+    // The actual TCP port number.
+    int port_;
+    // Syncronize accept handling on the same socket.
+    grpc_core::Mutex mu_;
+    // This will hold the socket for the next accept.
+    SOCKET accept_socket_ ABSL_GUARDED_BY(mu_) = INVALID_SOCKET;
+    // The listener winsocket.
+    grpc_core::OrphanablePtr<WinSocket> listener_socket_ ABSL_GUARDED_BY(mu_);
+    EventEngine::ResolvedAddress listener_sockname_;
+  };
+  absl::StatusOr<SinglePortSocketListener*> AddSinglePortSocketListener(
+      SOCKET sock, EventEngine::ResolvedAddress addr);
+
+  IOCP* iocp_;
```

Comment:
maybe change the ordering of the data members to start with the stuff that doesn't change through the life of the object.

Reply:
I like that, done.